### PR TITLE
Feature nxp 24034 separate trashed from lifecycle

### PR DESCRIPTION
**Pull request by kevinleturc**
https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-kleturc/388/

### CODE REVIEW

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/trash/test/AbstractTestTrashService.java [hunk 1]
```diff
@@ -225,101 +244,90 @@ public void testUndeleteChildren() throws Exception {
...
+    public void testTrashCheckedInDocumentDefault() {
         doTestTrashCheckedInDocument(true);
     }
 
-    protected void doTestTrashCheckedInDocument(boolean expectCheckedIn) throws Exception {
+    protected void doTestTrashCheckedInDocument(boolean expectCheckedIn) {
         DocumentModel folder = session.createDocumentModel("/", "folder", "Folder");
         folder = session.createDocument(folder);
-        DocumentModel doc = session.createDocumentModel("/folder", "doc", "File");
+        DocumentModel doc = session.createDocumentModel(folder.getPathAsString(), "doc", "File");
```
Comment (efge) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/trash/test/AbstractTestTrashService.java]: (detail) not really needed, a constant is shorter and more explicit.

File: nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/AbstractTrashService.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        ctx.setProperty(CoreEventConstants.SESSION_ID, session.getSessionId());
+        Event event = ctx.newEvent(eventId);
+        event.setInline(false);
+        event.setImmediate(true);
+        EventService eventService = Framework.getService(EventService.class);
+        eventService.fireEvent(event);
+    }
+
+    /**
+     * {@inheritDoc}
```
Comment (efge) [inline nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/AbstractTrashService.java]: Unneeded

File: nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+public class PropertyTrashService extends AbstractTrashService {
+
+    private static final Log log = LogFactory.getLog(PropertyTrashService.class);
+
+    public static final String SYSPROP_IS_TRASHED = "isTrashed";
+
+    @Override
+    public boolean isTrashed(CoreSession session, DocumentRef docRef) {
+        Boolean isTrashed = session.getDocumentSystemProp(docRef, SYSPROP_IS_TRASHED, Boolean.class);
+        return isTrashed != null && isTrashed.booleanValue();
```
Comment (efge) [inline nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java]: Simpler: `Boolean.TRUE.equals(isTrashed)`

File: nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/LifeCycleTrashService.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        session.save();
+    }
+
+    protected void trashDocument(CoreSession session, DocumentModel doc) {
+        if (doc.getParentRef() == null) {
+            // handle placeless document
+            session.removeDocument(doc.getRef());
+        } else {
+            if (!Boolean.parseBoolean(String.valueOf(doc.getContextData(DISABLE_TRASH_RENAMING)))) {
```
Comment (efge) [inline nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/LifeCycleTrashService.java]: You can compare directly, no need go through a String: `if (!Boolean.TRUE.equals(doc.getContextData(DISABLE_TRASH_RENAMING)))`

File: nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/LifeCycleTrashService.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                // stop if lifecycle properties can't be read on an ancestor
+                log.debug("Stopping to restore ancestors because " + ancestorRef.toString() + " is not readable");
+                break;
+            }
+        }
+    }
+
+    protected void undeleteDocument(CoreSession session, DocumentModel doc) {
+        String name = doc.getName();
+        if (!Boolean.parseBoolean(String.valueOf(doc.getContextData(DISABLE_TRASH_RENAMING)))) {
```
Comment (efge) [inline nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/LifeCycleTrashService.java]: Same comment

File: nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    protected Set<DocumentRef> doUntrashDocument(DocumentModel doc, boolean processChildren) {
+        CoreSession session = doc.getCoreSession();
+        DocumentRef docRef = doc.getRef();
+
+        // move only non placeless document
+        // currently we don't trash placeless document
+        DocumentRef parentRef = doc.getParentRef();
+        if (!Boolean.parseBoolean(String.valueOf(doc.getContextData(DISABLE_TRASH_RENAMING))) && parentRef != null) {
```
Comment (efge) [inline nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java]: Same comment

File: nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/TrashService.java [hunk 7]
```diff
@@ -33,20 +34,59 @@
...
+
+    /***
+     * Event for a document untrashed by the user. Triggers an async listener that untrashes its children too.
+     *
+     * @since 10.1
+     */
+    String DOCUMENT_UNTRASHED = "documentUntrashed";
+
+    /**
+     * Skip the renaming.
```
Comment (efge) [inline nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/TrashService.java]: Mention that this is a key for the context data, and that it happens if the value is `Boolean.TRUE`

**Comment by efge** [comment]
👍

**Comment by driverpt** [comment]
👍

**Comment by funshodavid** [comment]
:+1:

**Pull request merged**